### ISSUE TITLE
Enable compiling modules without library paths

### ISSUE THREAD

**Issue by bobbinth** [0xPolygonMiden/miden-vm#1071]
Current implementation of [CompileModule()](https://github.com/0xPolygonMiden/miden-vm/blob/next/assembly/src/assembler/mod.rs#L187) requires that we provide a full library module (i.e., both `ModuleAst` and `LibraryPath`). However, sometimes we may want to compile a module without a library path (e.g., for `AccountCode` in Miden Rollup). So, we should probably refactor `CompileModule` to look as follows:
```Rust
pub fn compile_module(
    &self,
    module: &ModuleAst,
    path: Option<&LibraryPath>,
    context: &mut AssemblyContext,
) -> Result<Vec<RpoDigest>, AssemblyError> {
```
Thus, tying a module to a library path would be optional. This has a few implications:
- First, we can create some "reserved" path for such modules - similar to how we do it for [kernel and executable modules](https://github.com/0xPolygonMiden/miden-vm/blob/next/assembly/src/library/path.rs#L30). I would probably name this `#anon`.
- Second, we won't be able to compute procedure IDs for procedures defined in such modules. This will affect how we handle re-exported and exported procedures. Specifically:
  - For re-exported procedures, we won't need to create additional alias mappings in the procedure cache.
  - For exported procedures, we'll need to add a procedure to procedure cache by its MAST root rather than ID.

For the last point, we could remove `id` field from `NamedProcedure` and move procedure ID computations out of `ModuleContext`. This would also imply that we'd change `insert()` method on `ProcedureCache` to look something like this:
```Rust
pub fn insert(&mut self, proc: NamedProcedure, id: Option<ProcedureId>) -> Result<(), AssemblyError>
```

**Comment by frisitano** [comment]
The client that is invoking the `compile_module(..)` method may want to know the `procedure name -> mast root` mapping such that they can identify the mast roots they are interested in easily.  I wonder if we should change the return type to be `Result<BTreeMap<ProcedureName, RpoDigest>, AssemblyError>`?

**Comment by bobbinth** [comment]
> I wonder if we should change the return type to be `Result<BTreeMap<ProcedureName, RpoDigest>, AssemblyError>`?

I was thinking about this as well. One minor difference is that I'd probably reverse the map to be `BTreeMap<RpoDigest, ProcedureName>`.

There is also another approach: we could add `exported_procs()` method to `ModuleAst`. Then, the clients would be able to iterate over both to get the mapping themselves.

**Comment by frisitano** [comment]
> I was thinking about this as well. One minor difference is that I'd probably reverse the map to be `BTreeMap<RpoDigest, ProcedureName>`.

What's your thinking behind using the `RpoDigest` as the key?  I think having the `ProcedureName` might be more practical as the user defines the `ProcedureName` and as such this will be entity the user is aware and can use to do lookups in the map. 

> There is also another approach: we could add `exported_procs()` method to `ModuleAst`. Then, the clients would be able to iterate over both to get the mapping themselves.

Could you elaborate on this?  How would the client map the MAST's tot he `ProcedureName`'s? My initial thinking is having the map as a return type from the `compile_module(..)` is a more natural fit.

**Comment by bobbinth** [comment]
> What's your thinking behind using the `RpoDigest` as the key? I think having the `ProcedureName` might be more practical as the user defines the `ProcedureName` and as such this will be entity the user is aware of and can use to do lookups in the map.

My thinking was that procedure hash is kind of like a unique identifier for the procedure, and thus lookups based on it could be more comment. For example, we may want to store account procedures in a database where the key is procedure hash, and value is a `ProcedureAst` (which contains procedure name). But I can see how using procedure name as the key could be useful as well.

> Could you elaborate on this? How would the client map the MAST's tot he `ProcedureName`'s? My initial thinking is having the map as a return type from the `compile_module(..)` is a more natural fit.

Assuming we have `exported_procs() -> impl Iterator<Item = &ProcedureAst>`  on `ModuleAst`, we could do something like this:
```Rust
let proc_hashes: Vec<RpoDigest> = assembler.compile_module(...);
for (proc_hash, proc_ast) in proch_hashes.iter().zip(module_ast.exported_proc()) {
    // build a map of proc_hash |-> proc_ast or a map proc_name |-> proc_hash etc.
}
```
This is a bit more complicated, but also a bit more flexible then returning a map.

**Comment by frisitano** [comment]
> My thinking was that procedure hash is kind of like a unique identifier for the procedure, and thus lookups based on it could be more comment. For example, we may want to store account procedures in a database where the key is procedure hash, and value is a `ProcedureAst` (which contains procedure name). But I can see how using procedure name as the key could be useful as well.

Yeah I think this makes sense.  What would be the purpose of such a database? My proposal was based of my experience working on the `create_note` auth PR in which I needed to fetch the MAST root for the `create_note` account procedure but future proofing this to support more use cases makes sense.  We could also consider introducing a `ModuleMetadata` struct with some simple helper functions:

```rust
pub struct ModuleMetadata {
    data: BTreeMap<RpoDigest, ProcedureAst>,
}

impl ModuleMetadata {
    pub fn get_mast_root(&self, proc_name: ProcedureName) -> RpoDigest {
        todo!()
    }

    ...

    pub fn inner(self) -> BTreeMap<RpoDigest, ProcedureAst> {
        self.data
    }
}
```

> Assuming we have `exported_procs() -> impl Iterator<Item = &ProcedureAst>` on `ModuleAst`, we could do something like this:
> 
> ```rust
> let proc_hashes: Vec<RpoDigest> = assembler.compile_module(...);
> for (proc_hash, proc_ast) in proch_hashes.iter().zip(module_ast.exported_proc()) {
>     // build a map of proc_hash |-> proc_ast or a map proc_name |-> proc_hash etc.
> }
> ```
> 
> This is a bit more complicated, but also a bit more flexible then returning a map.

I think abstracting this from the client by doing the mapping within the `compile_module(..)` method would be cleaner.  What additional flexibility does this solution provide over returning a map?

**Comment by bobbinth** [comment]
> What would be the purpose of such a database?

I was thinking for `miden-client` one of the tables could be store mappings from `RpoDigest` to `ProcedureAst` so that we don't have to store full `ModuleAst` with each instance of an account. But now thinking about this, this may be more tricky than I originally though.

> We could also consider introducing a ModuleMetadata struct with some simple helper functions

I'm not sure about this. This mapping may not always be interested to the user of this method and it is relatively "heavy" (the ASTs could be fairly big).

I've also just realized that returning `ProcedureAst` for all exported procedures would be difficult as we may not have access to ASTs of re-exported procedures. So, probably for now we should keep the interface fairly light (e.g., limit the return type to a light-weight vector or map).

Given the above, I think probably a simple map between names and digests would work best. I still have a slight preference for `BTreeMap<RpoDigest, ProcedureName>` vs. `BTreeMap<ProcedureName, RpoDigest>` - but could go either way.

**Comment by frisitano** [comment]
> I'm not sure about this. This mapping may not always be interested to the user of this method and it is relatively "heavy" (the ASTs could be fairly big).

Agreed it is probably too cumbersome.

> Given the above, I think probably a simple map between names and digests would work best. I still have a slight preference for `BTreeMap<RpoDigest, ProcedureName>` vs. `BTreeMap<ProcedureName, RpoDigest>` - but could go either way.

I do agree that `RpoDigest` is a more natural unique identifier for a procedure from a systems perspective, however I think we need to be use case driven here.  The primary use case I see for the returned map is to map from `ProcedureName`s which are user specified to `RpoDigest`'s. As such I think having `BTreeMap<ProcedureName, RpoDigest>` makes the most sense. In addition, there is no guarantee that a procedure MAST root will be unique, however there is a guarantee that the procedure name will be unique (as this is enforced by the parser) - as such we also mitigate concerns around MAST root collisions having the map configured this way.

**Comment by bobbinth** [comment]
One thing I realized, is that at some point (maybe relatively soon), we'll need to enable compilation of modules with "un-named" procedures. One use case for this is code that we put on-chain. For such code (both note scripts and account code), we'd probably want to strip procedure names to reduce the code size. Thus, if we want to parse/compile code which comes from on-chain accounts/notes, we won't have procedure names available. I'm not sure how we'd return a map in this case.

**Comment by Overcastan** [comment]
Closed by #1079.
